src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,400 @@
+            set
+            {
+                if (value == null || value.TotalMilliseconds <= 0)
+                {
+                    this.heartbeatInterval = TimeSpan.FromMilliseconds(DefaultHeartbeatIntervalMs);
+                }
+                else
+                {
+                    this.heartbeatInterval = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the currently defined instrumentation key to send heartbeat telemetry items to.
+        /// 
+        /// Note that if the heartbeat provider has not been initialized yet, this key would get reset to 
+        /// whatever the telemetry configuration the heartbeat provider is initialized with.
+        /// </summary>
+        public string InstrumentationKey
+        {
+            get => this.telemetryClient?.InstrumentationKey;
+            set
+            {
+                if (this.telemetryClient != null)
+                {
+                    this.telemetryClient.InstrumentationKey = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not heartbeats are enabled
+        /// </summary>
+        public bool IsEnabled
+        {
+            get => this.isEnabled;
+            set
+            {
+                if (!this.isEnabled && value)
+                {
+                    // we need to start calling the timer again
+                    this.HeartbeatTimer.Change(this.heartbeatInterval, this.heartbeatInterval);
+                }
+
+                this.isEnabled = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a list of default field names that should not be sent with each heartbeat.
+        /// </summary>
+        public IEnumerable<string> DisabledDefaultFields
+        {
+            get => this.disabledDefaultFields;
+            set
+            {
+                this.disabledDefaultFields = null;
+                if (value != null)
+                {
+                    this.disabledDefaultFields = new List<string>(value);
+                }
+
+                this.SetDefaultPayloadItems();
+            }
+        }
+
+        private Timer HeartbeatTimer { get; set; } // timer that will send each heartbeat in intervals
+
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            if (this.telemetryClient == null)
+            {
+                this.telemetryClient = new TelemetryClient(configuration);
+            }
+
+            this.SetDefaultPayloadItems();
+
+            this.isEnabled = true;
+
+            // Note: if this is a subsequent initialization, the interval between heartbeats will be updated in the next cycle so no .Change call necessary here
+            if (this.HeartbeatTimer == null)
+            {
+                this.HeartbeatTimer = new Timer(this.HeartbeatPulse, this, this.heartbeatInterval, this.heartbeatInterval);
```

Comment:
will this interval change in runtime be applied in runtime?

Reply:
Yes, if you look at the way we've implemented ‘HeartbeatPulse’:

                    this.HeartbeatTimer.Change(Timeout.Infinite, Timeout.Infinite);
                    if (this.isEnabled)
                    {
                        hp.Send();
                        this.HeartbeatTimer.Change(this.Interval, this.Interval);
                    }

…we disable the Timer setting the timeout to infinite, and then setting the Timer to the value of Interval. In this way, the timer will get updated with the new interval whenever the Timer ‘fires’ (or whenever the ‘IsEnabled’ property changes from false to true).